sql/src/test/java/io/crate/integrationtests/SeqNoBasedOCCIntegrationTest.java

```diff
@@ -0,0 +1,161 @@
+        assertEquals(1L, response.rowCount());
+        refresh();
+
+        execute("delete from t where x = 1 and _seq_no = ? and _primary_term = ?",
+                new Object[]{preUpdateSeqNo, preUpdatePrimaryTerm});
+        assertEquals(0, response.rowCount());
+    }
+
+    @Test
+    public void testUpdateWhereSeqNoAndPrimaryTermWithPrimaryKey() throws Exception {
+        execute("create table t (x integer primary key, y string)");
+        ensureYellow();
+
+        execute("insert into t (x, y) values (?, ?)", new Object[]{1, "don't panic"});
+        refresh();
+
+        execute("select _seq_no, _primary_term from t where x = 1");
+        assertEquals(1L, response.rowCount());
+        Long preUpdateSeqNo = (Long) response.rows()[0][0];
+        Long preUpdatePrimaryTerm = (Long) response.rows()[0][1];
+
+        execute("update t set y = ? where x = ? and _seq_no = ? and _primary_term = ?",
+                new Object[]{"ok now panic", 1, preUpdateSeqNo, preUpdatePrimaryTerm});
+        assertEquals(1L, response.rowCount());
+
+        // Validate that the row is really updated
+        refresh();
+        execute("select y from t where x = 1");
+        assertEquals(1L, response.rowCount());
+        assertEquals("ok now panic", response.rows()[0][0]);
+    }
+
+    @Test
+    public void testUpdateWhereSeqNoAndPrimaryTermWithoutPrimaryKey() throws Exception {
+        expectedException.expect(SQLActionException.class);
+        expectedException.expectMessage(VersioninigValidationException.SEQ_NO_AND_PRIMARY_TERM_USAGE_MSG);
+        execute("create table t (x integer primary key, y string) with (number_of_replicas=0)");
+        ensureYellow();
+        execute("update t set y = ? where _seq_no = ? and _primary_term = ?",
+                new Object[]{"ok now panic", 1, 5});
+    }
+
+    @Test
+    public void testUpdateWhereSeqNoAndPrimaryTermWithConflict() throws Exception {
+        execute("create table t (x integer primary key, y string)");
+        ensureYellow();
+
+        execute("insert into t (x, y) values (?, ?)", new Object[]{1, "don't panic"});
+        refresh();
+
+        execute("select _seq_no, _primary_term from t where x = 1");
+        assertEquals(1L, response.rowCount());
+        Long preUpdateSeqNo = (Long) response.rows()[0][0];
+        Long preUpdatePrimaryTerm = (Long) response.rows()[0][1];
+
+        execute("update t set y = ? where x = ? and _seq_no = ? and _primary_term = ?",
+                new Object[]{"hopefully not updated", 1, preUpdateSeqNo + 4, preUpdatePrimaryTerm});
+        assertEquals(0L, response.rowCount());
+
+        // Validate that the row is really NOT updated
+        refresh();
+        execute("select y from t where x = 1");
+        assertEquals(1L, response.rowCount());
+        assertEquals("don't panic", response.rows()[0][0]);
+    }
+
+    @Test
+    public void testSelectWhereSeqNoAndPTWithoutPrimaryKey() throws Exception {
```

Comment:
Can we turn this into a unit test ? (several hundred ms, vs a couple of ms runtime)